### PR TITLE
Improve error handling

### PR DESCRIPTION
**Pull request by rainerschoe**
- remove some `unwrap`, `panic` and `expect` uses
- Improve error messages
- rework segment evaluation to return errors instead of panic
- Add/rework testcases for some error scenarios

I experimented using `anyhow` to create error chains providing the user with more readable error messages.
For now I terminate `anyhow` usage in segment evaluation code and use the `thiserror` types for everything user-facing.
Feedback appreciated.

The behavior differs from the go implementation in strictness. I'd appreciate Input from @saikumar1607 on what is the intended behavior for error handling in entity evaluation. Is it fine to have the strict enforcement as implemented here?

### CODE REVIEW

File: src/segment_evaluation.rs [hunk 1]
```diff
@@ -19,125 +19,156 @@ use crate::{
...
     entity: &impl Entity,
-) -> Option<TargetingRule> {
+) -> Result<Option<TargetingRule>> {
     let mut targeting_rules = segment_rules.collect::<Vec<_>>();
     targeting_rules.sort_by(|a, b| a.order.cmp(&b.order));
-    targeting_rules
-        .into_iter()
-        .find(|targeting_rule| targeting_rule_applies_to_entity(segments, targeting_rule, entity))
+
+    for targeting_rule in targeting_rules.into_iter(){
```
Comment (rainerschoe) [inline src/segment_evaluation.rs:36]: I failed to find a nice way of propagating errors out of iterator chains. So in many cases like this I needed to fall-back to explicit loops. Any ideas?

File: src/segment_evaluation.rs [hunk 2]
```diff
@@ -19,125 +19,156 @@ use crate::{
...
+        if targeting_rule_applies_to_entity(segments, &targeting_rule, entity)
+        .map_err(
+            |e|{
+            // This terminates the use of anyhow in this module, converting all errors:
```
Comment (rainerschoe) [inline src/segment_evaluation.rs:40]: This terminates `anyhow` use, We might lose some information (stacktrace) here.

File: Cargo.toml [hunk 3]
```diff
@@ -15,6 +15,7 @@ tungstenite = { version = "0.24.0", features = ["native-tls"] }
 url = "2.5.2"
 http = "1.1.0"
 thiserror = "2.0.4"
+anyhow = "1.0.94"
```
Comment (jgsogo) [inline Cargo.toml:18]: We don't want anyhow as a dependency for the library

File: src/segment_evaluation.rs [hunk 4]
```diff
@@ -19,125 +19,156 @@ use crate::{
     entity::{AttrValue, Entity},
     models::TargetingRule,
 };
+use crate::errors::{Result, Error};
+
+// For chaining errors creating useful error messages:
+use anyhow::{Context, Result as AnyhowResult};
+use anyhow::anyhow;
```
Comment (jgsogo) [inline src/segment_evaluation.rs:26]: We don't really want `anyhow` in library code. It hides most of information about the actual error. Let's reserve `anyhow` for application.

File: src/client/feature.rs [hunk 5]
```diff
@@ -43,13 +43,13 @@ impl Feature {
                 Value::Numeric(NumericValue(model_value.0.clone()))
             }
             crate::models::ValueKind::Boolean => {
-                Value::Boolean(model_value.0.as_bool().ok_or(Error::ProtocolError)?)
+                Value::Boolean(model_value.0.as_bool().ok_or(Error::ProtocolError("Expected Boolean".into()))?)
```
Comment (jgsogo) [inline src/client/feature.rs:46]: If we want to, later we can add more typed information (`ErrorBooleanExpected`, `ErrorStringExpected`) so the message is just for humans... and we can also include what was the value we were trying to parse as boolean.

**Review by jgsogo — CHANGES_REQUESTED** [review]
I really think we want to avoid `anyhow` in library code. With `thiserror` we can propagate all the info to our users and they can use pattern matching to decide what to do with them.

File: src/segment_evaluation.rs (hunk 1 above)
Comment (jgsogo) [reply src/segment_evaluation.rs:36]: https://github.com/rainerschoe/appconfiguration-rust-sdk/pull/1

**Review by jgsogo — APPROVED** [review]

**Pull request merged**